backend/cfg/vectorize.ml

```diff
@@ -73,808 +231,2730 @@ end = struct
+          to_bits offset_in_bytes = (get_width_in_bits t1 |> Option.get)
+
+      (** [is_before t1 t2] returns true if we can prove that t1 and t2 are disjoint
+          intervals within the same block, and t1 is before t2. *)
+      let is_before ~arg_offset_in_bytes t1 t2 block =
+        let res =
+          match offset_in_bytes ~arg_offset_in_bytes t1 t2 with
+          | None -> false
+          | Some offset_in_bytes ->
+            State.dump_debug (Block.state block)
+              "offset_in_bytes = %d, get_width_in_bits t1 = %d \n"
+              (to_bits offset_in_bytes)
+              (get_width_in_bits t1 |> Option.get);
+            to_bits offset_in_bytes >= (get_width_in_bits t1 |> Option.get)
         in
-        let items_in_vector =
-          vector_width_in_bytes
-          / Memory_accesses.Memory_operation.width starting_memory_operation
+        State.dump_debug (Block.state block) "is_before (%a) (%a) = %b \n"
+          Instruction.print t1.instruction Instruction.print t2.instruction res;
+        res
+
+      (* [is_adjacent] implies [is_disjoint
+         ~arg_offset_in_bytes:actual_arg_offset_in_bytes] *)
+      let is_disjoint ~arg_offset_in_bytes t1 t2 block =
+        let res =
+          is_before ~arg_offset_in_bytes t1 t2 block
+          || is_before ~arg_offset_in_bytes t2 t1 block
         in
-        let can_cross_chunk seed instruction =
-          List.fold_left
-            (fun can memory_operation ->
-              can
-              && can_cross memory_accesses
-                   (Memory_accesses.Memory_operation.instruction
-                      memory_operation)
-                   instruction)
-            true seed
-        in
-        let rec find_seed n seed cell_option =
-          if n = 0
-          then Some seed
-          else
-            match cell_option with
-            | None -> None
-            | Some cell ->
-              let instruction = Instruction.Basic (DLL.value cell) in
-              if Instruction.is_store instruction
-              then
-                let new_store =
-                  Instruction.id instruction
-                  |> Memory_accesses.get_memory_operation_exn memory_accesses
-                in
-                if Memory_accesses.Memory_operation.is_adjacent (List.hd seed)
-                     new_store
-                then find_seed (n - 1) (new_store :: seed) (DLL.next cell)
-                else None
-              else if can_cross_chunk seed instruction
-              then find_seed n seed (DLL.next cell)
-              else None
-        in
-        find_seed (items_in_vector - 1)
-          [starting_memory_operation]
-          starting_cell
-        |> Option.map List.rev)
-      stores
-
-  let dump ppf (seeds : t list) =
+        State.dump_debug (Block.state block) "is_disjoint (%a) (%a) = %b \n"
+          Instruction.print t1.instruction Instruction.print t2.instruction res;
+        res
+
+      let maybe_closure_block t =
```

Comment:
The test for `Alloc` is only generated at `Emit`. I think that the middle end has enough information about types to know when expressions evaluate to closures or not. We can propagate it to the backend in the future and have more precise `Reg.typ` for example, instead of the current heuristic. Added a comment.